File: ui/src/lib/adminClients.ts
/**
 * Small fetch wrapper for the superuser admin Clients routes
 * (`/api/v1/admin/clients*`).
 *
 * These routes are not part of the generated OpenAPI client yet, so this
 * mirrors the client's base-URL resolution and Bearer-token convention
 * (same pattern as lib/kyc.ts).
 */

export type VoiceLinkLiveState =
  | "active"
  | "missing"
  | "unconfigured"
  | "unknown";

export interface AdminClient {
  organization_id: number;
  organization_name: string;
  owner_user_id?: number | null;
  owner_email?: string | null;
  owner_provider_id?: string | null;
  created_at?: string | null;
  voicelink_status?: string | null;
  voicelink_client_id?: string | null;
  voicelink_username?: string | null;
  voicelink_error?: string | null;
  has_voicelink_config: boolean;
  did_number?: string | null;
  // Live reconciliation against VoiceLink.
  live_state: VoiceLinkLiveState;
  live_client_id?: string | null;
  // Remaining call-seconds balance; null = unmetered (unlimited).
  credits_seconds_remaining?: number | null;
  // Billing/plan enrichment (added by the richer admin list endpoint). These
  // are optional so the row still renders if the backend has not shipped them.
  effective_plan?: string | null;
  // Remaining balance in INR; null = unmetered (unlimited).
  money_left_inr?: number | null;
  money_spent_inr?: number | null;
  per_minute_inr?: number | null;
  suspended?: boolean;
}

// The five sellable plans, in ascending order, for the plan selectors.
export const ADMIN_PLANS = [
  "trial",
  "starter",
  "growth",
  "scale",
  "enterprise",
] as const;
export type AdminPlan = (typeof ADMIN_PLANS)[number];

export interface AdminClientNote {
  at?: string | null;
  by?: string | number | null;
  text: string;
}

export interface AdminClientMoney {
  balance_seconds?: number | null;
  unlimited?: boolean | null;
  per_minute_inr?: number | null;
  money_left_inr?: number | null;
  spent_seconds?: number | null;
  money_spent_inr?: number | null;
}

export interface AdminClientPricing {
  per_minute_inr?: number | null;
  number_price_inr?: number | null;
  setup_fee_inr?: number | null;
  // Map of pricing-field name -> true when that field is a per-client override
  // (rather than derived from the plan default).
  custom?: Record<string, boolean> | null;
}

export interface AdminClientVoiceLink {
  status?: string | null;
  client_id?: string | null;
  username?: string | null;
  did_number?: string | null;
}

export interface AdminClientUsage {
  total_calls?: number | null;
  total_minutes?: number | null;
  [key: string]: number | null | undefined;
}

// Detail payload for GET /admin/clients/{orgId}. Every nested block is optional
// so the page keeps rendering if the parallel backend has not shipped a field.
export interface AdminClientDetail {
  organization_id: number;
  organization_name: string;
  owner_email?: string | null;
  owner_user_id?: number | null;
  owner_provider_id?: string | null;
  plan?: string | null;
  plan_override?: string | null;
  features?: { api?: boolean | null; mcp?: boolean | null } | null;
  pricing?: AdminClientPricing | null;
  money?: AdminClientMoney | null;
  suspended?: boolean;
  notes?: AdminClientNote[] | null;
  voicelink?: AdminClientVoiceLink | null;
  kyc?: AdminClientKycStatus | null;
  usage?: AdminClientUsage | null;
}

// PATCH /admin/clients/{orgId}/profile — every field optional; only sent fields
// are applied. A `null` on a pricing field clears the override (reset to plan
// default); a `null` on plan_override clears the plan override.
export interface AdminProfilePatch {
  plan_override?: string | null;
  per_minute_inr?: number | null;
  number_price_inr?: number | null;
  setup_fee_inr?: number | null;
  suspended?: boolean;
}

export interface ChargeSetupFeeResult {
  balance?: number | null;
  money?: AdminClientMoney | null;
}

export interface CreateAdminClientBody {
  email: string;
  name?: string;
  plan?: string;
  initial_credit_minutes?: number;
}

export interface CreateAdminClientResult {
  organization_id?: number;
  [key: string]: unknown;
}

export interface AdminAuditEntry {
  id: number;
  actor_user_id?: number | null;
  target_organization_id?: number | null;
  action: string;
  detail?: unknown;
  created_at?: string | null;
}

export interface AdminClientsListResult {
  clients: AdminClient[];
}

export interface RetryProvisionResult {
  voicelink_status: string;
  voicelink_client_id?: string | null;
  voicelink_username?: string | null;
  voicelink_error?: string | null;
}

export interface CreateClientResult {
  action: string; // "linked" | "created"
  voicelink_status: string;
  voicelink_client_id?: string | null;
  voicelink_username?: string | null;
  voicelink_error?: string | null;
}

export interface AssignDidBody {
  did_number: string;
  client_id?: string;
}

export interface AssignDidResult {
  configuration_id: number;
  created: boolean;
  did_number: string;
  client_id?: string | null;
}

export interface GrantCreditsResult {
  organization_id: number;
  granted_seconds: number;
  credits_seconds_remaining?: number | null;
}

// Backend detail string for GET /password when no display copy is stored
// (absent, encryption key unset, or the stored token failed to decrypt).
export const NO_STORED_PASSWORD = "no_stored_password";

export interface ClientPasswordResult {
  username?: string | null;
  password: string;
}

export interface RecordPasswordResult {
  organization_id: number;
  stored: boolean;
  // Reminder: this is a record of the portal password, not a change on
  // VoiceLink (there is no upstream change-password API).
  note: string;
}

// "ok" = fetched from VoiceLink | "no_client" = org has no VoiceLink client
// id | "disabled" = reseller credentials unset on the backend.
export type AdminKycState = "ok" | "no_client" | "disabled";

export interface AdminClientKycStatus {
  status: AdminKycState;
  enabled: boolean;
  client_id_configured?: boolean;
  has_voicelink_config?: boolean;
  client_id?: string | null;
  kyc_status?: string | null;
  pan_verified?: boolean | null;
  aadhaar_verified?: boolean | null;
  gst_verified?: boolean | null;
  is_complete?: boolean | null;
  current_step?: number | string | null;
  account_type?: string | null;
}

function backendUrl(): string {
  return (
    process.env.NEXT_PUBLIC_BACKEND_URL ||
    (typeof window !== "undefined" ? window.location.origin : "")
  );
}

function detailFromBody(body: unknown): string {
  const e = body as { detail?: unknown };
  if (typeof e?.detail === "string") return e.detail;
  if (Array.isArray(e?.detail) && e.detail.length > 0) {
    const first = e.detail[0] as { msg?: string };
    if (first?.msg) return first.msg;
  }
  return "Request failed";
}

async function adminRootFetch<T>(
  token: string,
  path: string,
  init?: RequestInit,
): Promise<T> {
  const res = await fetch(`${backendUrl()}/api/v1/admin${path}`, {
    ...init,
    headers: {
      "Content-Type": "application/json",
      Authorization: `Bearer ${token}`,
      ...(init?.headers ?? {}),
    },
  });
  let body: unknown = {};
  try {
    body = await res.json();
  } catch {
    // Non-JSON response body — fall through to the generic error below.
  }
  if (!res.ok) throw new Error(detailFromBody(body));
  return body as T;
}

// Convenience wrapper for the `/admin/clients` sub-tree (the majority of
// routes). `path` is appended after `/admin/clients` (e.g. "" or "/42/notes").
function adminFetch<T>(
  token: string,
  path: string,
  init?: RequestInit,
): Promise<T> {
  return adminRootFetch<T>(token, `/clients${path}`, init);
}

export const listAdminClients = (token: string) =>
  adminFetch<AdminClientsListResult>(token, "");

export const retryProvisionClient = (
  token: string,
  organizationId: number,
  password: string,
) =>
  adminFetch<RetryProvisionResult>(
    token,
    `/${organizationId}/retry-provision`,
    { method: "POST", body: JSON.stringify({ password }) },
  );

export const createClientForOrg = (
  token: string,
  organizationId: number,
  password?: string,
) =>
  adminFetch<CreateClientResult>(token, `/${organizationId}/create`, {
    method: "POST",
    body: JSON.stringify(password ? { password } : {}),
  });

export const assignDidToClient = (
  token: string,
  organizationId: number,
  body: AssignDidBody,
) =>
  adminFetch<AssignDidResult>(token, `/${organizationId}/assign-did`, {
    method: "POST",
    body: JSON.stringify(body),
  });

export const grantCreditsToClient = (
  token: string,
  organizationId: number,
  minutes: number,
) =>
  adminFetch<GrantCreditsResult>(token, `/${organizationId}/grant-credits`, {
    method: "POST",
    body: JSON.stringify({ minutes }),
  });

export const getClientKycStatus = (token: string, organizationId: number) =>
  adminFetch<AdminClientKycStatus>(token, `/${organizationId}/kyc-status`);

export const getClientPassword = (token: string, organizationId: number) =>
  adminFetch<ClientPasswordResult>(token, `/${organizationId}/password`);

export const recordClientPassword = (
  token: string,
  organizationId: number,
  password: string,
) =>
  adminFetch<RecordPasswordResult>(token, `/${organizationId}/password`, {
    method: "POST",
    body: JSON.stringify({ password }),
  });

export const getAdminClientDetail = (token: string, organizationId: number) =>
  adminFetch<AdminClientDetail>(token, `/${organizationId}`);

export const updateAdminProfile = (
  token: string,
  organizationId: number,
  patch: AdminProfilePatch,
) =>
  adminFetch<AdminClientDetail>(token, `/${organizationId}/profile`, {
    method: "PATCH",
    body: JSON.stringify(patch),
  });

export const addClientNote = async (
  token: string,
  organizationId: number,
  text: string,
): Promise<AdminClientNote[]> => {
  // The endpoint returns { organization_id, notes: [...] } — unwrap to the list.
  const res = await adminFetch<{ notes?: AdminClientNote[] }>(
    token,
    `/${organizationId}/notes`,
    { method: "POST", body: JSON.stringify({ text }) },
  );
  return Array.isArray(res?.notes) ? res.notes : [];
};

export const chargeSetupFee = (
  token: string,
  organizationId: number,
  amountInr?: number,
) =>
  adminFetch<ChargeSetupFeeResult>(token, `/${organizationId}/charge-setup-fee`, {
    method: "POST",
    body: JSON.stringify(amountInr != null ? { amount_inr: amountInr } : {}),
  });

export const createAdminClient = (token: string, body: CreateAdminClientBody) =>
  adminFetch<CreateAdminClientResult>(token, "", {
    method: "POST",
    body: JSON.stringify(body),
  });

export const listAdminAudit = async (
  token: string,
  organizationId?: number,
  limit = 50,
): Promise<AdminAuditEntry[]> => {
  const qs = new URLSearchParams();
  if (organizationId != null) qs.set("org_id", String(organizationId));
  qs.set("limit", String(limit));
  // The endpoint returns { items: [...] } — unwrap to the array the UI maps over.
  const res = await adminRootFetch<{ items?: AdminAuditEntry[] }>(
    token,
    `/audit?${qs.toString()}`,
  );
  return Array.isArray(res?.items) ? res.items : [];
};

File: api/schemas/admin_clients.py
"""Request/response schemas for the superuser admin Clients endpoints."""

from datetime import datetime
from typing import Dict, List, Optional

from pydantic import BaseModel, Field, field_validator

from api.schemas.kyc import KycStatusResponse
from api.services.plans import ASSIGNABLE_PLANS


class AdminClientItem(BaseModel):
    organization_id: int
    organization_name: str
    owner_user_id: Optional[int] = None
    owner_email: Optional[str] = None
    owner_provider_id: Optional[str] = None
    created_at: Optional[datetime] = None
    voicelink_status: Optional[str] = None
    voicelink_client_id: Optional[str] = None
    voicelink_username: Optional[str] = None
    voicelink_error: Optional[str] = None
    has_voicelink_config: bool = False
    did_number: Optional[str] = None
    # Live reconciliation against VoiceLink (GET /v1/reseller/clients):
    # "active" (exists in VoiceLink) | "missing" | "unconfigured" (reseller
    # creds unset) | "unknown" (reseller lookup failed → stored status shown).
    live_state: str = "unknown"
    live_client_id: Optional[str] = None
    # Remaining call-seconds balance; None = unmetered (unlimited).
    credits_seconds_remaining: Optional[int] = None
    # SaaS billing view (effective plan + money-in-INR at the client's rate).
    effective_plan: str = "trial"
    per_minute_inr: float = 0.0
    # Money remaining in INR; None = unmetered (unlimited).
    money_left_inr: Optional[float] = None
    money_spent_inr: float = 0.0
    suspended: bool = False


class AdminClientsListResponse(BaseModel):
    clients: List[AdminClientItem]


class RetryProvisionRequest(BaseModel):
    """A NEW VoiceLink password; provisioning keeps an encrypted display copy."""

    password: str

    @field_validator("password")
    @classmethod
    def password_min_length(cls, v: str) -> str:
        if len(v) < 8:
            raise ValueError("Password must be at least 8 characters")
        return v


class RetryProvisionResponse(BaseModel):
    voicelink_status: str
    voicelink_client_id: Optional[str] = None
    voicelink_username: Optional[str] = None
    voicelink_error: Optional[str] = None


class CreateClientRequest(BaseModel):
    """Optional password override for one-click create.

    Normally omitted — the endpoint reuses the org's stored (encrypted)
    signup password. A password is only supplied for legacy orgs that have no
    stored secret.
    """

    password: Optional[str] = None

    @field_validator("password")
    @classmethod
    def password_min_length(cls, v: Optional[str]) -> Optional[str]:
        if v is not None and len(v) < 8:
            raise ValueError("Password must be at least 8 characters")
        return v


class CreateClientResponse(BaseModel):
    action: str  # "linked" (already existed) | "created"
    voicelink_status: str
    voicelink_client_id: Optional[str] = None
    voicelink_username: Optional[str] = None
    voicelink_error: Optional[str] = None


class AssignDidRequest(BaseModel):
    did_number: str = Field(
        ...,
        min_length=1,
        description="DID in its VoiceLink-registered form (e.g. 919484959244)",
    )
    client_id: Optional[str] = Field(
        default=None,
        description=(
            "VoiceLink client id to stamp on the configuration; defaults to "
            "the org's provisioned voicelink_client_id"
        ),
    )
    arm_kyc: bool = Field(
        default=False,
        description=(
            "Also stamp voicelink_did_purchased_at so the KYC dialing gate "
            "applies to this org (marketplace purchases always do this)"
        ),
    )


class AssignDidResponse(BaseModel):
    configuration_id: int
    created: bool
    did_number: str
    client_id: Optional[str] = None


class GrantCreditsRequest(BaseModel):
    """Top-up for a metered org's call-credits balance (1 credit = 1 minute)."""

    minutes: int = Field(
        ...,
        ge=1,
        le=100_000,
        description="Minutes of call credit to grant (converted to seconds).",
    )


class GrantCreditsResponse(BaseModel):
    organization_id: int
    granted_seconds: int
    # Balance after the grant; never None here (unmetered orgs are rejected).
    credits_seconds_remaining: Optional[int] = None


class ClientPasswordResponse(BaseModel):
    """The stored display copy of the client's VoiceLink portal password."""

    username: Optional[str] = None
    password: str


class RecordPasswordRequest(BaseModel):
    """A portal password to record (encrypted at rest, display-only).

    This is our copy of the password the owner set in the VoiceLink portal —
    recording it does NOT change the password on VoiceLink (there is no
    upstream change-password API).
    """

    password: str = Field(..., min_length=8)


class RecordPasswordResponse(BaseModel):
    organization_id: int
    stored: bool
    # Reminder surfaced to the operator: this is a record of the portal
    # password, not a password change on VoiceLink.
    note: str


class AdminKycStatusResponse(KycStatusResponse):
    """Per-org KYC status for the admin Clients view.

    Same shape as the self-serve ``GET /kyc/status`` response plus a
    ``status`` discriminator:

    - ``ok`` — the org's VoiceLink ``client_id`` resolved and the KYC status
      was fetched from VoiceLink.
    - ``no_client`` — the org has no resolvable VoiceLink client id (KYC
      would act on the reseller's own account, so we don't fetch).
    - ``disabled`` — reseller credentials are not configured.
    """

    status: str = "ok"
    client_id: Optional[str] = None


# ======== Per-client profile / detail (SaaS admin) ========


class AdminPricingCustom(BaseModel):
    """Which pricing fields are per-client overrides (for a "custom" badge)."""

    per_minute_inr: bool = False
    number_price_inr: bool = False
    setup_fee_inr: bool = False


class AdminPricing(BaseModel):
    """Effective per-client pricing (INR), global defaults where not overridden."""

    per_minute_inr: float
    number_price_inr: int
    setup_fee_inr: int
    custom: AdminPricingCustom


class AdminMoney(BaseModel):
    """Money view: balance + spend in both seconds and INR at the client's rate."""

    balance_seconds: Optional[int] = None
    unlimited: bool = False
    per_minute_inr: float = 0.0
    money_left_inr: Optional[float] = None
    spent_seconds: int = 0
    money_spent_inr: float = 0.0


class AdminNote(BaseModel):
    """One entry in the org's admin ops log."""

    at: Optional[str] = None
    by: Optional[int] = None
    text: str = ""


class AdminClientUsage(BaseModel):
    """Rolling usage summary (from get_organization_overview)."""

    period: str = "month"
    total_calls: int = 0
    total_minutes: float = 0.0
    connected_calls: int = 0
    money_spent_inr: float = 0.0


class AdminClientDetailResponse(BaseModel):
    """Full per-client admin view (identity + plan + pricing + money + KYC)."""

    organization_id: int
    organization_name: str
    owner_user_id: Optional[int] = None
    owner_email: Optional[str] = None
    owner_provider_id: Optional[str] = None
    created_at: Optional[datetime] = None
    # VoiceLink state (stored — a lighter view than the list's live reconcile).
    voicelink_status: Optional[str] = None
    voicelink_client_id: Optional[str] = None
    voicelink_username: Optional[str] = None
    voicelink_error: Optional[str] = None
    has_voicelink_config: bool = False
    did_number: Optional[str] = None
    # Plan: ``plan`` is the effective tier (override wins); ``plan_override`` is
    # the raw admin override (null = derived from purchases).
    plan: str
    plan_override: Optional[str] = None
    features: Dict[str, bool] = Field(default_factory=dict)
    pricing: AdminPricing
    money: AdminMoney
    suspended: bool = False
    notes: List[AdminNote] = Field(default_factory=list)
    kyc: AdminKycStatusResponse
    # Omitted (null) if the usage rollup could not be computed.
    usage: Optional[AdminClientUsage] = None


class AdminProfileUpdateRequest(BaseModel):
    """Partial per-client profile update. Only *sent* fields change; send a
    pricing/plan field as ``null`` to clear the override back to the default."""

    plan_override: Optional[str] = None
    per_minute_inr: Optional[float] = Field(default=None, ge=0)
    number_price_inr: Optional[int] = Field(default=None, ge=0)
    setup_fee_inr: Optional[int] = Field(default=None, ge=0)
    suspended: Optional[bool] = None

    @field_validator("plan_override")
    @classmethod
    def validate_plan(cls, v: Optional[str]) -> Optional[str]:
        if v is not None and v not in ASSIGNABLE_PLANS:
            raise ValueError(
                f"plan_override must be one of {ASSIGNABLE_PLANS} or null to clear"
            )
        return v


class AdminProfileResponse(BaseModel):
    """The client's profile after an update: effective plan + pricing + flags."""

    organization_id: int
    plan: str
    plan_override: Optional[str] = None
    features: Dict[str, bool] = Field(default_factory=dict)
    pricing: AdminPricing
    suspended: bool = False


class AddNoteRequest(BaseModel):
    text: str = Field(..., min_length=1)


class AdminNotesResponse(BaseModel):
    organization_id: int
    notes: List[AdminNote] = Field(default_factory=list)


class ChargeSetupFeeRequest(BaseModel):
    """Optional override of the configured setup fee (INR)."""

    amount_inr: Optional[int] = Field(default=None, ge=1)


class ChargeSetupFeeResponse(BaseModel):
    organization_id: int
    fee_inr: int
    charged_seconds: int
    credits_seconds_remaining: Optional[int] = None
    money: AdminMoney


class CreateClientAccountRequest(BaseModel):
    """Create a brand-new client org + owner user."""

    email: str = Field(..., min_length=3)
    name: Optional[str] = None
    plan: Optional[str] = None
    initial_credit_minutes: Optional[int] = Field(default=None, ge=0, le=1_000_000)

    @field_validator("plan")
    @classmethod
    def validate_plan(cls, v: Optional[str]) -> Optional[str]:
        if v is not None and v not in ASSIGNABLE_PLANS:
            raise ValueError(f"plan must be one of {ASSIGNABLE_PLANS}")
        return v


class CreateClientAccountResponse(BaseModel):
    organization_id: int
    owner_user_id: int
    owner_email: str
    organization_name: str
    plan: str
    credits_seconds_remaining: Optional[int] = None
    # Generated login password for the new owner (superuser-only response) — the
    # owner should change it after first login.
    temporary_password: str
    note: str


class AdminAuditItem(BaseModel):
    id: int
    actor_user_id: Optional[int] = None
    target_organization_id: Optional[int] = None
    action: str
    detail: Optional[dict] = None
    created_at: Optional[datetime] = None


class AdminAuditListResponse(BaseModel):
    items: List[AdminAuditItem] = Field(default_factory=list)
